yo1k/tic_tac_toe/core/game.py

```diff
@@ -0,0 +1,282 @@
+
+    @staticmethod
+    def new_occupy(cell: Cell) -> Action:
+        return Action(False, cell, False)
+
+    @staticmethod
+    def new_start() -> Action:
+        return Action(False, None, True)
+
+    @property
+    def surrender(self):
+        return self.__surrender
+
+    @property
+    def occupy(self):
+        return self.__occupy
+
+    @property
+    def start(self):
+        return self.__start
+
+    def __str__(self):
+        if self.__surrender is True:
+            action = "surrender"
+        elif self.__occupy is not None:
+            action = f"occupy={self.__occupy}"
+        elif self.__start is True:
+            action = "start"
+        else:
+            assert False
+        return (
+                f"{type(self)}("
+                f"{action})")
+
+
+class ActionQueue(ABC):
+    @abstractmethod
+    def next(self) -> Optional[Action]:
+        pass
+
+
+class Logic:
+    """Game logic."""
+
+    def __init__(self, action_queues: Sequence[ActionQueue]):
+        """Indexes in `action_queues` correspond to indexes in `State.players`."""
+        assert len(action_queues) == State.players_count(), f"{action_queues}"
+        self.__action_queues = action_queues
+
+    def advance(self, state: State):
+        turn = state.turn()
+        action = self.__action_queues[turn].next()
+        if action is None:
+            return None
+        elif action.start is True:
+            Logic.__start(state)
+        elif action.surrender is True:
+            Logic.__surrender(state)
+        elif action.occupy is not None:
+            Logic.__occupy(state, action.occupy)
+        else:
+            assert False
+
+    @staticmethod
+    def __occupy(state: State, cell: Cell):
+        assert state.phase is Phase.INROUND
+        turn = state.turn()
+        mark = state.players[turn].mark
+        if state.board.get(cell) is None:
+            state.board.set(cell, mark)
+        else:
+            assert False
+        if Logic.win_condition(state.board, cell):
+            Logic.__win(state)
+        elif Logic.__last_step(state.step):
+            Logic.__draw(state)
+        else:
+            state.step += 1
+
+    @staticmethod
+    def win_condition(board: Board, last_occupied: Cell) -> bool:
+        h_match = 0
+        v_match = 0
+        d1_match = 0
+        d2_match = 0
+        x = last_occupied.x
+        y = last_occupied.y
+        mark = board.get(last_occupied)
+        for i in range(Board.size()):
+            if board.get(Cell(i, y)) == mark:
+                h_match += 1
+            if board.get(Cell(x, i)) == mark:
+                v_match += 1
+            if board.get((Cell(i, i))) == mark:
+                d1_match += 1
+            if board.get(Cell(i, Board.size() - 1 - i)) == mark:
+                d2_match += 1
+        return (
+                h_match == Board.size() or v_match == Board.size()
+                or d1_match == Board.size() or d2_match == Board.size())
```

Comment:
Does the following core make the linter unhappy?
```python
return (h_match == Board.size() or v_match == Board.size()
        or d1_match == Board.size() or d2_match == Board.size())
```
If not, this is how we should write it.

Reply:
Done.